README.md
# sweetsheet
Show beautiful bottom sheet as confirmation dialog quickly and easily.

|                                         nice                                         |                                          warning                                           |
| :----------------------------------------------------------------------------------: | :----------------------------------------------------------------------------------------: |
| ![nice](https://github.com/CorneilleEdi/sweetsheet/raw/master/assets/nice_sheet.jpg) | ![warning](https://github.com/CorneilleEdi/sweetsheet/raw/master/assets/warning_sheet.jpg) |


|                                       success                                        |                                       danger                                       |
| :----------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------: |
| ![success](https://github.com/CorneilleEdi/sweetsheet/raw/master/assets/success.jpg) | ![danger](https://github.com/CorneilleEdi/sweetsheet/raw/master/assets/danger.jpg) |

and since version `0.2.0` , it is fully customizable.

|                                    custom dark                                     |                                   custom light                                    |
| :--------------------------------------------------------------------------------: | :-------------------------------------------------------------------------------: |
| ![success](https://github.com/CorneilleEdi/sweetsheet/raw/master/assets/black.jpg) | ![danger](https://github.com/CorneilleEdi/sweetsheet/raw/master/assets/light.jpg) |


## Getting Started

With increasingly large smartphones, it has become almost impossible to have good ergonomics in an application using confirmation dialogs. 
The SweetSheet package allows you to show beautiful bottoms sheets to replace your confirmation dialogs and maintain a good user experience.

## How to use?

Add this to your package's pubspec.yaml file:

```
dependencies:
  sweetsheet: ^0.4.0
```

And import it:
```dart
import 'package:sweetsheet/sweetsheet.dart';
```

to start using the sheets, you have to create the instance of the 
`SweetSheet` class first.

```dart
final SweetSheet _sweetSheet = SweetSheet();
```

Now you can show the bottom sheet by calling the method `show()` on the instance.
This is the signature of the `show` method

```dart
  show({
    required BuildContext context,
    Text? title,
    required Text description,
    required CustomSheetColor color,
    required SweetSheetAction positive,
    EdgeInsets? contentPadding = const EdgeInsets.symmetric(horizontal: 24.0, vertical: 24.0),
    EdgeInsets? actionPadding = const EdgeInsets.all(8.0),
    SweetSheetAction? negative,
    IconData? icon,
    bool useRootNavigator = false,
    bool isDismissible = true,
  }) {}
```

This is the signature of the `SweetSheetAction` widget.

```dart
class SweetSheetAction extends StatelessWidget {
  final String title;
  final VoidCallback onPressed;
  final IconData icon;
  final Color color;

  SweetSheetAction({
    @required this.title,
    @required this.onPressed,
    this.icon,
    this.color = Colors.white,
  });  
}
```

> Attention : At least one action is required and the icons are IconData for uniformity purpose.


The `SweetSheetColor` is a class which have four static values and it determine the color of the sheet by creating a `CustomSheetColor`:

- SweetSheetColor.SUCCESS (green)
- SweetSheetColor.DANGER (red)
- SweetSheetColor.WARNING (orange)
- SweetSheetColor.NICE (blue)

## Customisation
To customise, create a new instance of `CustomSheetColor` and pass it to the argument `color` when calling the method show.
signature of `CustomSheetColor`:

```dart
class CustomSheetColor {
  Color main;
  Color accent;
  Color icon;

  CustomSheetColor({@required this.main, @required this.accent, this.icon});
}
```
To customise the action, pass a new color to `color` argument of `SweetSheetAction`.

## Example:

#### Warning

```dart
_sweetSheet.show(
    context: context,
    title: Text("Attention"),
    description: Text(
        'Your app is not connected to internet actually, please turn on Wifi/Celullar data.'),
    color: SweetSheetColor.WARNING,
    icon: Icons.portable_wifi_off,
    positive: SweetSheetAction(
      onPressed: () {
        Navigator.of(context).pop();
      },
      title: 'OPEN SETTING',
      icon: Icons.open_in_new,
    ),
    negative: SweetSheetAction(
      onPressed: () {
        Navigator.of(context).pop();
      },
      title: 'CANCEL',
    ),
);
```

#### Custom Light

```dart
_sweetSheet.show(
    context: context,
    description: Text(
      'Place your order. Please confirm the placement of your order : Iphone X 128GB',
      style: TextStyle(color: Color(0xff2D3748)),
    ),
    color: CustomSheetColor(
      main: Colors.white,
      accent: Color(0xff5A67D8),
      icon: Color(0xff5A67D8),
    ),
    icon: Icons.local_shipping,
    positive: SweetSheetAction(
      onPressed: () {
        Navigator.of(context).pop();
      },
      title: 'CONTINUE',
    ),
    negative: SweetSheetAction(
      onPressed: () {
        Navigator.of(context).pop();
      },
      title: 'CANCEL',
  ),
);
```

### That is all. Have fun!


CHANGELOG.md
# 0.1.0

- first release
  
# 0.1.1

- Change `SweetSheetAction` icon type to IconData

# 0.1.2

- Add more uniformity

# 0.1.3

- Force white color for contents to prevent color changing by custom ThemeData

# 0.2.0

- **Breaking Change** Change title and description to `Text`
- **Breaking Change** Change `type` argument to `color`- make title optional
- add color to icon
- add color to action
- add `CustomSheetColor` for Colors customisation
  
# 0.3.0

- **Breaking Change** Migrate to null safety
- Allow for a root navigator (issue [#3](https://github.com/CorneilleEdi/sweetsheet/issues/3))

# 0.3.1

- Add `isDismissible` to prevent clicking off the alert or dragging to dismiss

# 0.4.0

- Add `contentPadding` and `actionPadding`


lib/src/sheets.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

class CustomSheetColor {
  Color main;
  Color accent;
  Color icon;

  CustomSheetColor(
      {required this.main, required this.accent, required this.icon});
}

class SweetSheetColor {
  static CustomSheetColor DANGER = CustomSheetColor(
    main: Color(0xffEF5350),
    accent: Color(0xffD32F2F),
    icon: Colors.white,
  );
  static CustomSheetColor SUCCESS = CustomSheetColor(
    main: Color(0xff009688),
    accent: Color(0xff00695C),
    icon: Colors.white,
  );
  static CustomSheetColor WARNING = CustomSheetColor(
    main: Color(0xffFF8C00),
    accent: Color(0xffF55932),
    icon: Colors.white,
  );
  static CustomSheetColor NICE = CustomSheetColor(
    main: Color(0xff2979FF),
    accent: Color(0xff0D47A1),
    icon: Colors.white,
  );
}

class SweetSheet {
  show({
    required BuildContext context,
    Text? title,
    required Text description,
    required CustomSheetColor color,
    required SweetSheetAction positive,
    EdgeInsets? contentPadding = const EdgeInsets.symmetric(horizontal: 24.0, vertical: 24.0),
    EdgeInsets? actionPadding = const EdgeInsets.all(8.0),
    SweetSheetAction? negative,
    IconData? icon,
    bool useRootNavigator = false,
    bool isDismissible = true,
  }) {
    showModalBottomSheet(
      isDismissible: isDismissible,
      enableDrag: isDismissible,
      context: context,
      isScrollControlled: true,
      useRootNavigator: useRootNavigator,
      builder: (BuildContext context) {
        return Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Container(
              width: double.infinity,
              color: color.main,
              padding: contentPadding,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  title == null
                      ? Container()
                      : DefaultTextStyle(
                          style: TextStyle(
                            fontSize: 24,
                            color: Colors.white,
                            fontWeight: FontWeight.bold,
                          ),
                          textAlign: TextAlign.start,
                          child: title),
                  _buildContent(color, description, icon)
                ],
              ),
            ),
            Container(
              padding: actionPadding,
              color: color.accent,
              child: Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: _buildActions(positive, negative),
              ),
            )
          ],
        );
      },
    );
  }

  _buildContent(CustomSheetColor color, Text description, IconData? icon) {
    return Padding(
      padding: const EdgeInsets.only(top: 12.0),
      child: SingleChildScrollView(
        child: icon != null
            ? Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                crossAxisAlignment: CrossAxisAlignment.center,
                children: <Widget>[
                  Expanded(
                    child: DefaultTextStyle(
                        style: TextStyle(
                            fontSize: 18,
                            color: Colors.white,
                            fontFamily: 'circular'),
                        child: description),
                  ),
                  SizedBox(
                    width: 16,
                  ),
                  Icon(
                    icon,
                    size: 52,
                    color: color.icon,
                  )
                ],
              )
            : DefaultTextStyle(
                style: TextStyle(
                  fontSize: 18,
                  color: Colors.white,
                ),
                child: description,
              ),
      ),
    );
  }

  _buildActions(SweetSheetAction? positive, SweetSheetAction? negative) {
    List<SweetSheetAction> actions = [];

    // This order is important
    // It helps to place the positive at the right and the negative before
    if (negative != null) {
      actions.add(negative);
    }

    if (positive != null) {
      actions.add(positive);
    }

    return actions;
  }
}

class SweetSheetAction extends StatelessWidget {
  final String title;
  final VoidCallback onPressed;
  final IconData? icon;
  final Color color;

  SweetSheetAction({
    required this.title,
    required this.onPressed,
    this.icon,
    this.color = Colors.white,
  });

  @override
  Widget build(BuildContext context) {
    return icon == null
        ? TextButton(
            onPressed: onPressed,
            child: Text(
              title,
              style: TextStyle(
                color: color,
              ),
            ),
          )
        : TextButton.icon(
            onPressed: onPressed,
            label: Text(
              title,
              style: TextStyle(
                color: color,
              ),
            ),
            icon: Icon(
              icon,
              color: color,
            ),
          );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sweetsheet/sweetsheet.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'SweetSheet',
      theme: ThemeData(
        brightness: Brightness.dark,
        primarySwatch: Colors.blue,
      ),
      home: HomePage(title: 'SweetSheet'),
    );
  }
}

class HomePage extends StatefulWidget {
  final String title;

  HomePage({required this.title});

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final SweetSheet _sweetSheet = SweetSheet();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        padding: const EdgeInsets.all(16.0),
        child: ListView(
          children: <Widget>[
            LargeButton(
              context: context,
              text: 'Success sheet',
              onClick: () {
                _sweetSheet.show(
                  context: context,
                  title: Text("Lorem Ipsum"),
                  description: Text(
                      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. no you condimentum finibus ut ut lorem. Ut pellentesque mauris ut arcu rutrum, at tincidunt arcu tincidunt"),
                  color: SweetSheetColor.SUCCESS,
                  positive: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),
            LargeButton(
              context: context,
              text: 'Danger sheet with icon not dismissible',
              onClick: () {
                _sweetSheet.show(
                  isDismissible: false,
                  context: context,
                  title: Text("Delete this post?"),
                  description:
                      Text("This action will permanently delete this post."),
                  color: SweetSheetColor.DANGER,
                  icon: Icons.delete,
                  positive: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'DELETE',
                  ),
                  negative: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),
            LargeButton(
              context: context,
              text: 'Warning sheet with icon',
              onClick: () {
                _sweetSheet.show(
                  context: context,
                  title: Text("Attention"),
                  description: Text(
                      'Your app is not connected to internet actually, please turn on Wifi/Celullar data.'),
                  color: SweetSheetColor.WARNING,
                  icon: Icons.portable_wifi_off,
                  positive: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'OPEN SETTING',
                    icon: Icons.open_in_new,
                  ),
                  negative: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),
            LargeButton(
              context: context,
              text: "Nice sheet with icon",
              onClick: () {
                _sweetSheet.show(
                  context: context,
                  title: Text("Connect your watch"),
                  description: Text(
                      'To import your health data, you have to connect your smartwatch fist.'),
                  color: SweetSheetColor.NICE,
                  icon: Icons.watch,
                  positive: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CONNECT',
                    icon: Icons.open_in_new,
                  ),
                  negative: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),
            LargeButton(
              context: context,
              text: "Sheet with custom dark color",
              onClick: () {
                _sweetSheet.show(
                  context: context,
                  title: Text("Welcome back"),
                  description: Text(
                      'We are happy to see you again. We have some new things for you.'),
                  color: CustomSheetColor(
                    main: Colors.black12,
                    accent: Colors.black87,
                    icon: Colors.white,
                  ),
                  icon: Icons.bubble_chart,
                  positive: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'LET\'S GO',
                  ),
                );
              },
            ),
            LargeButton(
              context: context,
              text: "Sheet with custom light color",
              onClick: () {
                _sweetSheet.show(
                  context: context,
                  description: Text(
                    'Place your order. Please confirm the placement of your order : Iphone X 128GB',
                    style: TextStyle(color: Color(0xff2D3748)),
                  ),
                  color: CustomSheetColor(
                    main: Colors.white,
                    accent: Color(0xff5A67D8),
                    icon: Color(0xff5A67D8),
                  ),
                  icon: Icons.local_shipping,
                  positive: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CONTINUE',
                  ),
                  negative: SweetSheetAction(
                    onPressed: () {
                      Navigator.of(context).pop();
                    },
                    title: 'CANCEL',
                  ),
                );
              },
            ),
          ],
        ),
      ),
    );
  }
}

class LargeButton extends StatelessWidget {
  final BuildContext context;
  final String text;
  final VoidCallback onClick;

  const LargeButton({
    Key? key,
    required this.context,
    required this.text,
    required this.onClick,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.only(bottom: 8.0),
      child: FlatButton(
        padding: const EdgeInsets.all(30),
        color: Colors.black45,
        onPressed: onClick,
        child: Container(
          width: double.infinity,
          child: Text(
            text,
            textAlign: TextAlign.center,
          ),
        ),
      ),
    );
  }
}
